text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,679 @@
+`setComponentTemplate` on the subclass would take precedence over the
+superclass.
+
+### Template Scope
+
+In this context, _scope_ refers to the set of names that can be used to refer
+to values, helpers, and other components from within your template. For
+example, when you type `{{@firstName}}`, `{{this.count}}`, or `<UserAvatar
+/>`, how does Ember (or even you, as the programmer) figure out what each of these
+refers to? The answer to that is determined by the template's scope.
+
+Today, the names available inside a particular template are determined by a
+few different rules:
+
+1. `this` always refers to the component instance (if the template has a backing class).
+2. Arguments (like `@firstName`) are placed into scope when you invoke the component.
+3. Block parameters (like `item` in `{{#each @items as |item|}}`) are in scope, but only inside the block.
+4. Anything else must be looked up dynamically through Ember's container.
+   This lookup process is somewhat complex and not always as fast as we'd like.
+
+Because the rules of how components, helpers, and other values are looked up
+can be confusing for new learners, we'd like to explore alternate APIs that
+allow users to explicitly import them and refer to them from templates.
+
+Today, it's not easy for addons to add additional names to a template's scope
+that take precedence over dynamic resolution.
+
+This RFC proposes that the `precompile` function, described earlier, also
+accepts a `scope` option that specifies additional identifiers (as strings)
+that are available in the template's scope:
+
+```js
+import { precompile } from '@ember/template-compiler';
+
+const json = precompile(`{{t "Hello!"}} <User @user={{this.currentuser}} />`, {
+  scope: ['User', 't']  
+});
+```
+
+The `scope` option is an array of zero or more strings. In this example, the
+identifiers `User` and `t` are added to the template scope. Identifiers
+provided this way are in scope for the entire template (unless shadowed by a
+block argument).
+
+This RFC further proposes that the `createTemplateFactory` function,
+described earlier, also accepts a `scope` option that returns the reified
+scope values at runtime:
+
+```js
+import { createTemplateFactory } from '@ember/template-factory';
+import User from './OtherComponent';
+import t from '../helpers/i18n/t';
+
+const json = /* wire format JSON here */
+const template = createTemplateFactory(json, {
+  scope: () => [User, t]
+});
+```
+
+Note that, unlike with `precompile`, the `scope` option here is not an array
+but a function that _returns_ an array. This allows the array to be created
+after the JavaScript module has finished evaluating, ensuring the bindings
+included in the array have had time to be fully initialized.
+
+The order of the array is unimportant, other than that the same order must be
+preserved between the identifiers passed to `precompile` and the
+corresponding values passed to `createTemplateFactory`.
```

Comment:
This reads super risky to me. The order at two "independent" locations must be the same. Couldn't it be mapped by name?

Reply:
I don't think it matters later on either way, once the scope is correctly associated, either by name or order, you should be able to pull the values off the object/array and use them in a minifiable way. The issue is with the `createTemplateFactory()` in the first place - creating that object and assigning the keys must be shipped to the client, and _that_ part cannot be minified.

If there are a fair number of imports per file, say 5, this could add up in large apps. It's not a _massive_ amount, yes, and if this were an API expected to be used by your average user I would totally agree, but the idea here is that this will be _compiled output_ from a preprocessor, so it's much less likely that it'll get messed up IMO